File: src/ros2_bridge/mappers/sensor_msgs/point_cloud2.rs
//! Typed mapper for `sensor_msgs/msg/PointCloud2`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn point_cloud2_to_bus(msg: ros_env::sensor_msgs::msg::PointCloud2) -> crate::sensor_msgs::msg::v1::PointCloud2 {
    crate::sensor_msgs::msg::v1::PointCloud2 {
        header: Some(crate::ros2_bridge::mappers::std_msgs::header::header_to_bus(msg.header)),
        height: msg.height,
        width: msg.width,
        fields: msg.fields.into_iter().map(crate::ros2_bridge::mappers::sensor_msgs::point_field::point_field_to_bus).collect(),
        is_bigendian: msg.is_bigendian,
        point_step: msg.point_step,
        row_step: msg.row_step,
        data: crate::ros2_bridge::mappers::convert::IntoU8Vec::into_u8_vec(msg.data),
        is_dense: msg.is_dense,
    }
}

pub(crate) fn point_cloud2_to_ros(bus: crate::sensor_msgs::msg::v1::PointCloud2) -> ros_env::sensor_msgs::msg::PointCloud2 {
    ros_env::sensor_msgs::msg::PointCloud2 {
        header: crate::ros2_bridge::mappers::std_msgs::header::header_to_ros(bus.header.unwrap_or_default()),
        height: bus.height,
        width: bus.width,
        fields: bus.fields.into_iter().map(crate::ros2_bridge::mappers::sensor_msgs::point_field::point_field_to_ros).collect(),
        is_bigendian: bus.is_bigendian,
        point_step: bus.point_step,
        row_step: bus.row_step,
        data: crate::ros2_bridge::mappers::convert::FromByteSeq::from_byte_seq(bus.data),
        is_dense: bus.is_dense,
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct SensorMsgsPointCloud2Mapper;

impl TypedTopicMapper for SensorMsgsPointCloud2Mapper {
    type Ros = ros_env::sensor_msgs::msg::PointCloud2;
    type Bus = crate::sensor_msgs::msg::v1::PointCloud2;

    fn type_name(&self) -> &'static str {
        "sensor_msgs/msg/PointCloud2"
    }

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(point_cloud2_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(point_cloud2_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/sensor_msgs/point_cloud2.py
"""Generated mapper for `sensor_msgs/msg/PointCloud2`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert
from robot_bus.ros2_bridge.mappers.std_msgs.header import header_to_bus, header_to_ros
from robot_bus.ros2_bridge.mappers.sensor_msgs.point_field import point_field_to_bus, point_field_to_ros

_BusMsg = None


def _bus_cls():
    global _BusMsg
    if _BusMsg is None:
        from robot_bus.sensor_msgs.msg.v1 import PointCloud2 as BusMsg

        _BusMsg = BusMsg
    return _BusMsg


def point_cloud2_to_bus(msg):
    BusMsg = _bus_cls()
    bus = BusMsg()
    bus.header.CopyFrom(header_to_bus(msg.header))
    bus.height = msg.height
    bus.width = msg.width
    bus.fields.extend([point_field_to_bus(x) for x in msg.fields])
    bus.is_bigendian = msg.is_bigendian
    bus.point_step = msg.point_step
    bus.row_step = msg.row_step
    bus.data = bytes(msg.data)
    bus.is_dense = msg.is_dense
    return bus


def point_cloud2_to_ros(bus):
    from sensor_msgs.msg import PointCloud2 as RosMsg

    out = RosMsg()
    out.header = header_to_ros(bus.header)
    out.height = bus.height
    out.width = bus.width
    out.fields = [point_field_to_ros(x) for x in bus.fields]
    out.is_bigendian = bus.is_bigendian
    out.point_step = bus.point_step
    out.row_step = bus.row_step
    out.data = bytes(bus.data)
    out.is_dense = bus.is_dense
    return out


class SensorMsgsPointCloud2Mapper:
    _ros_type = None

    def ros_msg_type(self):
        cls = type(self)
        if cls._ros_type is None:
            from sensor_msgs.msg import PointCloud2 as RosMsg

            cls._ros_type = RosMsg
        return cls._ros_type

    def ros_to_bus(self, msg) -> bytes:
        return point_cloud2_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        BusMsg = _bus_cls()
        bus = BusMsg()
        bus.ParseFromString(payload)
        return point_cloud2_to_ros(bus)
